### PR TITLE
feat(devices): add vxlan device adapter

### PR DESCRIPTION
**Pull request by GeorgyKirichenko**
- Add a vxlan device adapter modeled on the vlan device: the input handler decapsulates the outer Ethernet/IPv4/UDP/VXLAN header chain and continues with the inner frame, and the output handler encapsulates with the configured VNI and underlay addresses, DF set on the outer IPv4 header, and a flow-hashed UDP source port.
- Drop non-matching, truncated, and fragmented outer packets on input, matching the decap module's tunnel-termination semantics, and enforce IPv4 datagram length limits on output.
- Ship the C dataplane and control-plane api library, the Go control plane with gRPC service and pre-FFI validation, a Rust CLI, web UI pages, and the shared-memory configuration flow.
- Register the device across meson, the control-plane bundle, the cargo workspace, the Makefile CLI list, and debian packaging.
- Cover decap, encap, malformed-input, and boundary cases with C unit tests and functional dataplane tests exercising the real handlers end to end.

### CODE REVIEW

File: devices/vxlan/web/device.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            { label: 'Dst IP', value: dstIp || '—', mono: true },
+        ];
+    },
+    createDefaults: () => ({
+        vni: 0,
+        dstPort: DEFAULT_DST_PORT,
+        srcMac: '',
+        dstMac: '',
+        srcIp: '',
+        dstIp: '',
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/web/device.ts:89]: Fixed: the manifest now provides editable settings via a loadDetail component (VxlanDetail.tsx, the trafgen pattern) covering all six fields, with validation mirroring the control plane. Covered by VxlanDetail.test.tsx plus a browser run asserting the emitted UpdateDevice body.

File: devices/vxlan/dataplane/dataplane.c [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		// A fragment must not be decapsulated: a non-first fragment's
+		// payload can mimic the UDP and VXLAN headers, and an MF-set
+		// first fragment ends mid-tunnel-packet.
+		if (ipv4_hdr->fragment_offset & rte_cpu_to_be_16(0x3FFF)) {
+			packet_front_drop(packet_front, packet);
+			continue;
+		}
+
+		offset += ihl * RTE_IPV4_IHL_MULTIPLIER;
+		if (data_len < offset + sizeof(struct rte_udp_hdr)) {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:127]: Fixed: decap now validates the declared envelopes before any mutation (total_length >= ihl*4, 14+total_length <= pkt_len, dgram_len >= UDP+VXLAN, udp_off+dgram_len <= ip_end) and trims trailing bytes to ip_end on accept (atomic rte_pktmbuf_trim, drop on failure). Gated by three new malformed-table cases and TestVxlan_DecapTrimsTrailingPadding; mutation-verified.

File: devices/vxlan/dataplane/dataplane.c [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+		);
+		if (vxlan_hdr->flags != VXLAN_FLAGS_BYTE ||
+		    rte_be_to_cpu_32(vxlan_hdr->vx_vni) >> 8 != config->vni) {
+			packet_front_drop(packet_front, packet);
+			continue;
+		}
+
+		offset += sizeof(struct rte_vxlan_hdr);
+		rte_pktmbuf_adj(mbuf, offset);
+		packet_refresh_data_len(packet);
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:194]: Fixed: the accept path now requires the inner Ethernet header to be resident in the head segment (data_len >= vxlan_end + 14), restoring parity with the parser's headers-in-head-segment contract. Gated by TestVxlan_DecapDropsNonResidentInnerHeader; mutation-verified.

File: devices/vxlan/dataplane/dataplane.c [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+		cp_device
+	);
+
+	struct packet *packet;
+	while ((packet = packet_list_pop(&packet_front->input)) != NULL) {
+		struct rte_mbuf *mbuf = packet_to_mbuf(packet);
+
+		// DF forbids fragmentation, so a frame that does not fit the
+		// 16-bit total length cannot be encapsulated at all.
+		if (rte_pktmbuf_pkt_len(mbuf) > UINT16_MAX - VXLAN_ENCAP_LEN) {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c]: Fixed: the threshold now excludes the outer Ethernet header (drop only above 65535 - 36); inner frames up to 65499 encapsulate with total_length 65535. Gated by TestVxlan_EncapJumboBoundary; mutation-verified.

File: devices/vxlan/dataplane/dataplane.c [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+			sizeof(struct rte_ipv4_hdr)
+		);
+		udp_hdr->dgram_cksum = 0;
+
+		struct rte_vxlan_hdr *vxlan_hdr =
+			(struct rte_vxlan_hdr *)(udp_hdr + 1);
+		vxlan_hdr->vx_flags = RTE_BE32(VXLAN_FLAGS_BYTE << 24);
+		vxlan_hdr->vx_vni = rte_cpu_to_be_32(config->vni << 8);
+
+		packet_front_output(packet_front, packet);
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:340]: Fixed: encap now rebases the parse metadata for the outer frame (network/transport header type+offset, flags, fragment_offset), mirroring the packet_decap idiom, without re-running parse_packet so the demux hash stays inner-derived. Gated behaviorally by TestVxlan_EncapRebasesParseMetadata (underlay-matched forward stage, blackhole fallthrough); mutation-verified.

**Comment by GeorgyKirichenko** [comment]
@codex review

**Comment by GeorgyKirichenko** [comment]
@codex review

File: devices/vxlan/dataplane/dataplane.c [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+			packet_front_drop(packet_front, packet);
+			continue;
+		}
+
+		// The declared UDP length must reach past the VXLAN header and
+		// stay inside the IPv4 datagram.
+		uint32_t dgram_len = rte_be_to_cpu_16(udp_hdr->dgram_len);
+		if (dgram_len < sizeof(struct rte_udp_hdr) +
+					sizeof(struct rte_vxlan_hdr) ||
+		    offset + dgram_len > ip_end) {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c]: Fixed: the trim target is now the declared UDP end (udp_off + dgram_len) rather than the IPv4 end, so a datagram declaring only the UDP+VXLAN headers can no longer inject the bytes beyond its payload. Gated by TestVxlan_DecapDropsInnerBeyondUdpDatagram and TestVxlan_DecapTrimsToUdpDatagramEnd; mutation-verified.

File: devices/vxlan/dataplane/dataplane.c [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+		 * fragment state clears.
+		 */
+		packet->network_header.type =
+			rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV4);
+		packet->network_header.offset = sizeof(struct rte_ether_hdr);
+		packet->transport_header.type = IPPROTO_UDP;
+		packet->transport_header.offset = sizeof(struct rte_ether_hdr) +
+						  sizeof(struct rte_ipv4_hdr);
+		packet->flags = 0;
+		packet->fragment_offset = 0;
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:287]: Fixed: the encap metadata rebase now resets the VLAN field to the untagged value (0) alongside the other outer-header fields, with a comment explaining why an explicit reset is needed although parse_packet would not clear it. Gated by TestVxlan_EncapResetsVlanMetadata via an untagged-only forward stage on the output pipeline; mutation-verified.

File: devices/vxlan/web/device.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            vni: ext.vni,
+            dst_port: ext.dstPort,
+            src_mac: ext.srcMac,
+            dst_mac: ext.dstMac,
+            src_ip: ext.srcIp,
+            dst_ip: ext.dstIp,
+        };
+    },
+    loadDetail: () => import('./VxlanDetail'),
+    save,
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/web/device.ts:117]: Fixed: the service now exposes a GetDevice RPC serving the last-applied settings from its update cache (clone-isolated, NotFound for unknown names), the web API gained the matching client call, and the manifest hydrates existing devices via loadData (trafgen pattern, graceful fallback to required-field flags on a cache miss). A pipeline-only edit now resubmits the fetched tunnel settings; covered by service round-trip tests and browser-verified.

File: devices/vxlan/dataplane/dataplane.c [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+			continue;
+		}
+
+		offset += sizeof(struct rte_vxlan_hdr);
+
+		// parse_packet bounds its header reads by the whole-packet
+		// length but reads through the head segment, so the inner
+		// Ethernet header must stay resident there after the adj — the
+		// same residency every RX packet arrives with.
+		if (data_len < offset + sizeof(struct rte_ether_hdr)) {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:173]: Answered with evidence, no change: the deeper split is the standard parser's own contract, not a device-specific gap. parse_ether_header and every subsequent parse_* helper bound by whole-packet length while reading head-segment memory (lib/dataplane/packet/packet.c), so a chained RX frame whose headers do not fit the head segment is already dropped-or-misread at the ingress parse for every port and module, vxlan or not. The added guard restores exactly that parity for the decap reparse by requiring the inner Ethernet header to be resident; requiring every parseable header would reject frames the ingress parser itself accepts. Tightening this belongs in the shared parser, not in this device.

File: devices/vxlan/dataplane/dataplane.c [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+			continue;
+		}
+
+		rte_pktmbuf_adj(mbuf, offset);
+		packet_refresh_data_len(packet);
+
+		// Re-run the standard parse so downstream pipelines see the
+		// inner frame's headers, hash and length; a malformed inner
+		// frame is dropped.
+		if (parse_packet(packet)) {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:206]: Fixed: the decap path now clears transport_header before the reparse, so a non-IP inner (ARP) no longer carries the outer UDP type and offset. Since no Go-loadable module exposes transport matching for non-IP-typed packets, the regression case lives in the permanent C suite (run_decap_transport_reset_test, seeded via the real RX parse call); mutation-verified (stale {17, 34} observed).

File: devices/vxlan/web/VxlanDetail.tsx [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+import { resolvedExt, DEFAULT_DST_PORT } from './types';
+
+// VNI is a 24-bit field and the outer UDP port a 16-bit one; the control
+// plane rejects anything outside these ranges, as it does MACs and IPs that
+// fail to parse, so the flags below name exactly what Save would reject.
+const VNI_MAX = 0xFFFFFF;
+const DST_PORT_MIN = 1;
+const DST_PORT_MAX = 0xFFFF;
+const DIGITS_RE = /^\d+$/;
+const MAC_RE = /^(?:[0-9a-fA-F]{2}[:-]){5}[0-9a-fA-F]{2}$/;
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/web/VxlanDetail.tsx]: Fixed: the regex is colon-only now, matching what the control plane accepts unmixed and what GetDevice serves back; mixed separators are flagged with the format hint. Covered in VxlanDetail.test.tsx.

**Comment by GeorgyKirichenko** [comment]
@codex review

**Comment by GeorgyKirichenko** [comment]
@codex review

File: devices/vxlan/controlplane/service.go [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+		DstIp:   m.settings.DstIP.String(),
+	}
+}
+
+// NewDeviceVxlanService creates a service publishing devices through the
+// given agent.
+func NewDeviceVxlanService(agent *ffi.Agent) *DeviceVxlanService {
+	return &DeviceVxlanService{
+		agent:   agent,
+		configs: map[string]*cachedDevice{},
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/controlplane/service.go:66]: Answered with evidence, no change in this PR: the process-local cache is the established device-service family design. The only other device with a read RPC (trafgen ShowConfig) has identical restart semantics — its cache is also empty after a control-plane restart — and vlan/plain expose no read at all. No read-back mechanism exists anywhere in the tree: the C device apis publish configs to shared memory one-way, so reconstructing the cache would require a new cross-cutting read-back API per device, out of scope here. The failure mode is safe and visible: NotFound surfaces as required-field flags in the UI, and a save without re-entered settings is rejected by validation, never silently applied. The intended state remains available in the pipeline operator's declarative config, which is the reconciliation source of truth in this architecture. A shared-memory read-back for device settings is worth a follow-up issue covering the whole family.

File: devices/vxlan/web/device.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        return {
+            vni: response.vni,
+            dstPort: response.dst_port,
+            srcMac: response.src_mac,
+            dstMac: response.dst_mac,
+            srcIp: response.src_ip,
+            dstIp: response.dst_ip,
+        };
+    } catch {
+        return {};
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/web/device.ts]: Fixed: loadData now suppresses only ApiError status 404 (the service's NotFound maps uniquely to 404; a registry miss is 503) and rethrows everything else, so the device stays unloaded rather than being blank-marked. Covered by new vitest cases asserting 404, 503, and non-ApiError propagation.

File: devices/vxlan/web/VxlanDetail.tsx [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    </div>
+);
+
+/** Editors for the six tunnel tunables; edits flow straight into the ext slice.
+ *
+ * Text fields bind the ext directly. The two numeric fields keep a local
+ * string so intermediate keystrokes stay editable, pushing a parsed value
+ * into the ext only while it is valid — Save always sends a parseable one.
+ */
+const VxlanDetail = ({ device, onUpdateExt }: DeviceDetailProps): React.JSX.Element => {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/web/VxlanDetail.tsx:68]: Fixed: the editors are deferred behind a loading state until the device is loaded (new devices are loaded at creation and render immediately), so edits can no longer be silently discarded by a late hydration response. Hook order is stable across both render paths; covered by a deferral test that reproduces the core's hydration merge.

**Comment by GeorgyKirichenko** [comment]
@codex review

File: devices/vxlan/dataplane/dataplane.c [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+		// Re-run the standard parse so downstream pipelines see the
+		// inner frame's headers, hash and length; a malformed inner
+		// frame is dropped. parse_packet sets network_header for
+		// every frame but leaves transport_header untouched for a
+		// non-IP one (e.g. ARP), which must not keep the outer UDP
+		// type and offset, so reset it to the fresh-packet state.
+		packet->transport_header.type = PACKET_HEADER_TYPE_UNKNOWN;
+		packet->transport_header.offset = 0;
+		if (parse_packet(packet)) {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:206]: Fixed: decap now resets the vlan id to the fresh-packet state alongside the transport metadata before the reparse, so a stripped outer VLAN tag can no longer reach the inner pipelines filters for an untagged inner frame. Gated by run_decap_vlan_reset_test in the vxlan dataplane test; mutation-verified.

File: devices/vxlan/web/VxlanDetail.tsx [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+        committed: number,
+        key: 'vni' | 'dstPort',
+        raw: string,
+    ): void => {
+        setInput(raw);
+        const trimmed = raw.trim();
+        if (DIGITS_RE.test(trimmed)) {
+            const parsed = Number(trimmed);
+            if (parsed >= min && parsed <= max && parsed !== committed) {
+                onUpdateExt({ [key]: parsed });
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/web/VxlanDetail.tsx]: Fixed: the numeric editors now bind the raw text straight into the ext slice (the local last-valid state is gone), the manifest gains extValid, and the Save button stays disabled until every field parses inside its range or matches the wire format; save itself refuses an invalid ext, so a programmatic caller cannot submit a stale last-valid number either. Covered by the rewritten vitest case walking an invalid VNI, a zero port, a bad Src IP and empty MACs through the gate to a clean save.

File: devices/vxlan/controlplane/service.go [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+	ctx context.Context,
+	request *vxlanpb.UpdateDeviceVxlanRequest,
+) (*vxlanpb.UpdateDeviceVxlanResponse, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	name := request.GetName()
+	if name == "" {
+		return nil, status.Error(codes.InvalidArgument, "module config name is required")
+	}
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/controlplane/service.go:82]: Fixed: UpdateDevice now rejects names longer than CP_DEVICE_NAME_LEN - 1 before any shared-memory call; the limit is taken from the C header via cgo (DeviceNameMaxLength), so the two cannot drift. Gated by a new invalid-argument case and TestUpdateDevice_NameLengthBoundary, which asserts the 79-byte name publishes and round-trips through GetDevice while the 80-byte one is rejected.

**Comment by GeorgyKirichenko** [comment]
@codex review

File: devices/vxlan/api/controlplane.c [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+	cp_device_vxlan->src_ip = settings->src_ip;
+	cp_device_vxlan->dst_ip = settings->dst_ip;
+
+	return &cp_device_vxlan->cp_device;
+}
+
+void
+cp_device_vxlan_free(struct cp_device *cp_device) {
+	struct agent *agent = ADDR_OF(&cp_device->agent);
+	cp_device_fini(cp_device);
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/api/controlplane.c]: Fixed: the device is now released through the reference tracker — cp_device_vxlan_free drops the construction reference via cp_device_release, a zero-refcount device parks on the agent instead of being torn down under a live reader, and cp_device_vxlan_new drains this type parked entries (typed destroy handler) before allocating, mirroring the vlan and trafgen lifecycle. Gated by the C dataplane test asserting a parked device is reclaimed whole by the next construction (mutation-verified against the drain) and by the Go drain test updated to the parking semantics.

File: devices/vxlan/controlplane/service.go [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+	// A longer name would be truncated into the fixed C name buffer, so
+	// the cache entry keyed by the full name and the published dataplane
+	// device would disagree, and two long names sharing the truncated
+	// prefix would overwrite the same device.
+	if len(name) > DeviceNameMaxLength {
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/controlplane/service.go]: Fixed: UpdateDevice now rejects names containing an embedded NUL alongside the length limit, before any shared-memory call. Gated by a new name_with_embedded_nul invalid-argument case.

File: devices/vxlan/dataplane/dataplane.c [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+		// encapsulated at all.
+		if (rte_pktmbuf_pkt_len(mbuf) >
+		    UINT16_MAX -
+			    (VXLAN_ENCAP_LEN - sizeof(struct rte_ether_hdr))) {
+			packet_front_drop(packet_front, packet);
+			continue;
+		}
+
+		struct rte_ether_hdr *ether_hdr = (struct rte_ether_hdr *)
+			rte_pktmbuf_prepend(mbuf, VXLAN_ENCAP_LEN);
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/dataplane/dataplane.c:264]: Fixed: encap now bounds the head segment separately (data_len must stay ≤ 65535 − 50 after the prepend), since rte_pktmbuf_prepend guards headroom only; a single-segment 65486–65499-byte frame is dropped instead of wrapping the 16-bit head length, while chained frames with small heads still encapsulate. Gated by run_encap_head_overflow_drop_test on the largest accepted frame; mutation-verified.

File: devices/vxlan/cli/src/main.rs [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+    pub name: String,
+    /// Pipeline assignments in format "pipeline_name:weight"
+    #[arg(short, long)]
+    pub input: Vec<String>,
+    /// Pipeline assignments in format "pipeline_name:weight"
+    #[arg(short, long)]
+    pub output: Vec<String>,
+    /// VXLAN network identifier
+    #[arg(short, long)]
+    pub vni: u32,
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply devices/vxlan/cli/src/main.rs]: Fixed: the VNI argument keeps only the long form --vni; the short -v stays owned by the global verbose flag. Verified with a clean cargo build and clippy of the crate.

**Comment by GeorgyKirichenko** [comment]
@codex review

**Comment by GeorgyKirichenko** [comment]
@codex review

File: controlplane/bundle/bundle.go [hunk 22]
```diff
@@ -189,6 +190,13 @@ func buildServices(
 				return trafgen.NewTrafgenDevice(devicesCfg.Trafgen.Unwrap(), trafgen.WithLog(log))
 			},
 		},
+		{
+			Name:       "vxlan device",
+			Configured: devicesCfg.Vxlan.Unwrap() != nil,
```
[the remark this answers is not in the document]
Comment (GeorgyKirichenko) [reply controlplane/bundle/bundle.go:195]: Fixed: the packaged controlplane.d/default.yaml now lists devices.vxlan (instance 0, shared memory_path, 16MB) beside plain and vlan, so the director registers DeviceVxlanService in packaged installs. Gated by Test_ShippedDefaultConfig_LoadsIntendedEnabledSet and the endpoint-defaults test, both extended to pin the enabled set including vxlan.

The input-handler semantics are unchanged and pinned by TestVxlan_InputDropsMalformedFrames: only unfragmented outer IPv4/UDP with the configured port, I flag and VNI is decapsulated — an MF-set first fragment, an offset-only later fragment, IPv6 outer, non-UDP protocol, wrong port and wrong VNI are all dropped unchanged, never routed past the device.
